### PR TITLE
Lift restrictions of access check for inlinalbe package symbols referenced in interfaces.

### PR DESCRIPTION
**Pull request by elsh**
Package decls are only printed in interface files if they are inlinable
(@usableFromInline, @inlinable, @_alwaysEmitIntoClient). They could be
referenced by a module outside of its defining module that belong to the same
package determined by the `package-name` flag. However, the flag is only in
.swiftmodule and .private.swiftinterface, thus type checking references of
inlinable package symbols in public interfaces fails due to the missing flag.
Instead of adding the package-name flag to the public interfaces, which
could raise a security concern, this PR grants access to such cases.

Resolves rdar://116142791

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tshortli** [comment]
I think we should strive to hide package names from public `.swiftinterfaces`. What if we just carved out a typechecking exception for `@usableFromInline package` decls in `.swiftinterfaces` instead, allowing them to be present without knowing the package name?

**Comment by elsh** [comment]
> I think we should strive to hide package names from public `.swiftinterfaces`. What if we just carved out a typechecking exception for `@usableFromInline package` decls in `.swiftinterfaces` instead, allowing them to be present without knowing the package name?

That was my initial thought but after having discussed more with @xymus this route seemed simpler (so users don't have to figure out why they sometimes see package-name and sometimes don't). I'm open to adding more restrictions though.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tshortli** [comment]
I don't think we should be too concerned about users being confused about why `-package-name` is absent from a public `.swiftinterface`; most developers have no reason to inspect a `.swiftinterface` and ultimately they are a compiler implementation detail. The typechecking rule that requires `-package-name` be specified when using `package` decls is meant to help developers avoid mistakenly forgetting to configure their package/project properly for `package` use. That's useful when typechecking source but I don't think the diagnostic makes as much sense in the context of typechecking a swiftinterface. I'd rather we complicate typechecking rules to treat swiftinterfaces differently (we do this a lot already) so that we can avoid leaking unnecessary details in the SDK.

**Review by zoecarver — APPROVED** [review]
Yay! Thank you Ellie!

**Comment by tshortli** [comment]
If what I'm suggesting is more complicated than I'm imagining to implement, then a version of this approach that would satisfy me more would be to print a placeholder package name instead of the real package name (ideally the placeholder value is known to the compiler and only accepted when parsing interfaces) .

**Comment by elsh** [comment]
Modified to print the flag only if the interface contains inlinable (@usableFromInline, @inlinalbe, @_alwaysEmitIntoClient) package decls.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tshortli** [comment]
> Modified to print the flag only if the interface contains inlinable (@usableFromInline, @inlinalbe, @_alwaysEmitIntoClient) package decls.

That's an improvement over always printing the package name, but I'd still like to know if just lifting the restriction entirely is a reasonable option so that we never have to print the package name `.swiftinterfaces` which is what I'd prefer.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -3827,6 +3827,18 @@ bool ValueDecl::isUsableFromInline() const {
...
+bool ValueDecl::skipAccessCheckIfInterface(const DeclContext *useDC,
+                                           AccessLevel useAcl,
+                                           AccessScope declScope) const {
+  if (!useDC || useAcl != AccessLevel::Package || !declScope.isPackage() ||
+      !isUsableFromInline() ||
+      getDeclContext()->getParentModule()->getBaseIdentifier() ==
+          useDC->getParentModule()->getBaseIdentifier())
+    return false;
+  auto useSF = useDC->getParentSourceFile();
+  return useSF && !useSF->isScriptMode();
```
Comment (tshortli) [inline lib/AST/Decl.cpp:3839]: Why does this check for `isScriptMode()`? I would expect this line to be 
```
return useSF && useSF->getKind() == SourceFileKind::Interface;
```

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -3827,6 +3827,18 @@ bool ValueDecl::isUsableFromInline() const {
   return false;
 }
 
+bool ValueDecl::skipAccessCheckIfInterface(const DeclContext *useDC,
+                                           AccessLevel useAcl,
+                                           AccessScope declScope) const {
+  if (!useDC || useAcl != AccessLevel::Package || !declScope.isPackage() ||
+      !isUsableFromInline() ||
+      getDeclContext()->getParentModule()->getBaseIdentifier() ==
```
Comment (tshortli) [inline lib/AST/Decl.cpp:3835]: This check determining whether the decl and the use are in the same module, right? I think I usually see that done by doing a pointer comparison of the `ModuleDecl *` values. Is there a reason the identifier strings need to be compared?

**Review by tshortli — COMMENTED** [review]
Thanks for implementing the approach I suggested! Looks like it turned out reasonably straightforward.

**Comment by elsh** [comment]
@swift-ci smoke test

**Review by tshortli — APPROVED** [review]

**Comment by elsh** [comment]
> Thanks for implementing the approach I suggested! Looks like it turned out reasonably straightforward.

Np! Thanks for suggesting and sticking with the approach so we can keep internal details hidden from users.

**Comment by elsh** [comment]
@swift-ci smoke test windows

**Pull request merged**